### PR TITLE
Add AVX512 operations

### PR DESCRIPTION
**Pull request by dong0321**
Status : Checked all operations with all types.

### CODE REVIEW

File: ompi/mca/op/intel_avx_op/op_avx_component.c [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        .opc_op_query = avx_component_op_query,
+    },
+};
+
+/*
+ * Component open
+ */
+static int avx_component_open(void)
+{
+    opal_output(ompi_op_base_framework.framework_output, "avx component open");
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_component.c]: remove.

File: ompi/mca/op/intel_avx_op/op_avx_component.c [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    /* A first level check to see if avx is even available in this
+       process.  E.g., you may want to do a first-order check to see
+       if hardware is available.  If so, return OMPI_SUCCESS.  If not,
+       return anything other than OMPI_SUCCESS and the component will
+       silently be ignored.
+
+       Note that if this function returns non-OMPI_SUCCESS, then this
+       component won't even be shown in ompi_info output (which is
+       probably not what you want).
+    */
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_component.c]: Where is the check ?

File: ompi/mca/op/intel_avx_op/op_avx_component.c [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+    return OMPI_SUCCESS;
+}
+
+/*
+ * Component close
+ */
+static int avx_component_close(void)
+{
+    opal_output(ompi_op_base_framework.framework_output, "avx component close");
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_component.c]: In fact remove all opal_output that are for debugging.

File: ompi/mca/op/intel_avx_op/op_avx_component.c [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+       resources that may have been allocated on this component).
+       Note that _component_close() will always be called at the end
+       of the process, so it may have been after any/all of the other
+       component functions have been invoked (and possibly even after
+       modules have been created and/or destroyed). */
+
+    return OMPI_SUCCESS;
+}
+
+static char *avx_component_version;
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_component.c]: What is this ?

File: ompi/mca/op/intel_avx_op/op_avx_component.c [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+    opal_output(ompi_op_base_framework.framework_output, "avx component register");
+
+    /* Additionally, since this component is simulating hardware,
+       let's make MCA params that determine whethere a) the hardware
+       is available, and b) whether double precision floating point
+       types are supported.  This allows you to change the behavior of
+       this component at run-time (by setting these MCA params at
+       run-time), simulating different kinds of hardware. */
+    mca_op_avx_component.hardware_available = true;
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_component.c]: This is not the final version right ?

File: ompi/mca/op/intel_avx_op/op_avx_component.c [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        }
+        break;
+    case OMPI_OP_BASE_FORTRAN_BXOR:
+        opal_output(ompi_op_base_framework.framework_output, "avx component op pick BXOR");
+        for (i = 0; i < OMPI_OP_BASE_TYPE_MAX; ++i) {
+            module->opm_fns[i] = ompi_op_avx_functions[OMPI_OP_BASE_FORTRAN_BXOR][i];
+            module->opm_3buff_fns[i] = ompi_op_avx_3buff_functions[OMPI_OP_BASE_FORTRAN_BXOR][i];
+        }
+        break;
+    case OMPI_OP_BASE_FORTRAN_LAND:
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_component.c]: No logical AND ?

File: ompi/mca/op/intel_avx_op/op_avx_functions.c [hunk 7]
```diff
@@ -0,0 +1,688 @@
...
+ *              sum, for integer 8,16,32,64
+ *
+ */
+#define OP_AVX_FUNC(name, type_sign, type_size, type, op) \
+    static void ompi_op_avx_2buff_##name##_##type(void *in, void *out, int *count, \
+            struct ompi_datatype_t **dtype, \
+            struct ompi_op_base_module_1_0_0_t *module) \
+{                                                                      \
+    int step;                                                          \
+    switch(type_size) {                                                \
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_functions.c]: I'm sure there is a better way to decide the step.

File: ompi/mca/op/intel_avx_op/op_avx_functions.c [hunk 8]
```diff
@@ -0,0 +1,688 @@
...
+        case 32: \
+                 step = 16; \
+        break; \
+        case 64: \
+                 step = 8;  \
+    }\
+    int size = *count/step; \
+    int i; \
+    int round = size*64; \
+    for (i = 0; i < round; i+=64) { \
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_functions.c]: Are these operations supposed to apply on cache-line aligned elements ?

File: ompi/mca/op/intel_avx_op/op_avx_functions.c [hunk 9]
```diff
@@ -0,0 +1,688 @@
...
+ *  Support ops: or, xor, and of 512 bits (representing integer data)
+ *
+ */
+#define OP_AVX_BIT_FUNC(name, type_size, type, op) \
+    static void ompi_op_avx_2buff_##name##_##type(void *in, void *out, int *count, \
+            struct ompi_datatype_t **dtype, \
+            struct ompi_op_base_module_1_0_0_t *module) \
+{                                                                      \
+    int step;                                                          \
+    switch(type_size) {                                                \
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_functions.c]: get this out and make a macro.

File: ompi/mca/op/intel_avx_op/op_avx_functions.c [hunk 10]
```diff
@@ -0,0 +1,688 @@
...
+        case 32: \
+                 step = 16; \
+        break; \
+        case 64: \
+                 step = 8;  \
+    }\
+    int size = *count/step; \
+    int i; \
+    int round = size*64; \
+    for (i = 0; i < round; i+=64) { \
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_functions.c]: This should also be a macro.

File: ompi/mca/op/intel_avx_op/op_avx_component.c [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+    return OMPI_SUCCESS;
+}
+
+/*
+ * Query whether this component wants to be used in this process.
+ */
+static int avx_component_init_query(bool enable_progress_threads,
+                                        bool enable_mpi_thread_multiple)
+{
+    if (mca_op_avx_component.hardware_available && !enable_mpi_thread_multiple) {
```
Comment (bosilca) [inline ompi/mca/op/intel_avx_op/op_avx_component.c]: Why are you disabling the component if threading is enabled ?

File: ompi/mca/op/avx/op_avx_functions.c [hunk 12]
```diff
@@ -77,22 +75,20 @@
     int size = *count/step; \
     int i; \
     int round = size*64; \
-    type* in = (type*)_in; \
-    type* out = (type*)_out; \
     for (i = 0; i < round; i+=64) { \
-        __m512i vecA =  _mm512_loadu_si512((in+i));\
-        __m512i vecB =  _mm512_loadu_si512((out+i));\
+        __m512i vecA =  _mm512_loadu_si512((_in+i));\
```
Comment (dong0321) [inline ompi/mca/op/avx/op_avx_functions.c]: Definition for _mm512_load_si512: 
 __m512i _mm512_load_si512 (void const* mem_addr)

I am not sure we need to do type convert. The types_per_step will be different based on the type, I tried to calculate the correct step for each type, but it seems not working. 

For correctness purpose I will remove type convert now. If we really need to specific type. I will fix later.

Comment (bosilca) [reply ompi/mca/op/avx/op_avx_functions.c]: This is plain ugly and highly non portable. I suggest you put back the typecast and add not 64 to i but ```step```. In fact you should model this on the float version below, without using the mask for the last set of operations but instead falling back to a lower vector op and then a Duff device.

**Comment by bosilca** [comment]
This has been moved to OMPI master in https://github.com/open-mpi/ompi/pull/7419/files

**Pull request closed without merging**